### PR TITLE
Support CXSMILES/CXSMARTS for reactions

### PR DESCRIPTION
**Pull request by greglandrum**
Since the CXSMILES extensions come at the end and refer to the atoms and bonds with continuously increasing indices, this required some reworking of the parsers.

### CODE REVIEW

File: Code/GraphMol/ChemReactions/DaylightParser.cpp [hunk 1]
```diff
@@ -106,75 +109,122 @@ ROMol *constructMolFromString(const std::string &txt,
...
 
 }  // end of namespace DaylightParserUtils
 
+namespace {
+void removeSpacesAround(std::string &text, size_t pos) {
+  auto nextp = pos + 1;
+  while (nextp < text.size() && (text[nextp] == ' ' || text[nextp] == '\t')) {
+    text.erase(nextp, 1);
+  }
+  if (pos > 0) {
```
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/DaylightParser.cpp:126]: Trying with c++ iterators, I don't think it simplifies the code much

```
auto nextp = trim.begin() + pos + 1;
if(pos + 1 < trim.size() && (*nextp == ' ' || *nextp == '\t')) {
 trim.erase(nextp, trim.find_first_not_of(" \t", nextp));
}
```

Comment (greglandrum) [reply Code/GraphMol/ChemReactions/DaylightParser.cpp:126]: agreed

File: Code/GraphMol/ChemReactions/DaylightParser.cpp [hunk 2]
```diff
@@ -106,75 +109,122 @@ ROMol *constructMolFromString(const std::string &txt,
...
+  std::string cxPart;
+  if (allowCXSMILES) {
+    auto sidx = origText.find_first_of("|");
+    if (sidx != std::string::npos && sidx != 0) {
+      text = origText.substr(0, sidx);
+      cxPart = boost::trim_copy(origText.substr(sidx, origText.size() - sidx));
+    }
+  }
+  // remove any spaces at the beginning, end, or before the '>'s
+  text = boost::trim_copy(text);
```
Comment (ricrogz) [inline Code/GraphMol/ChemReactions/DaylightParser.cpp:153]: You could just do `boost::trim(text)` here, and save the copy.

File: Code/GraphMol/ChemReactions/DaylightParser.cpp [hunk 3]
```diff
@@ -106,75 +109,122 @@ ROMol *constructMolFromString(const std::string &txt,
...
+
+  // re-find the '>' characters so that we can split on them
+  pos.clear();
+  for (std::size_t i = 0; i < text.length(); ++i) {
+    if (text[i] == '>' && (i == 0 || text[i - 1] != '-')) {
+      pos.push_back(i);
+    }
+  }
 
+  // there's always the chance that one or more of the "<" was in the name
```
Comment (ricrogz) [inline Code/GraphMol/ChemReactions/DaylightParser.cpp:196]: This is ">", right?

**Review by ricrogz — APPROVED** [review]
LGTM, just a couple of minor comments.

**Pull request merged**